### PR TITLE
refactor: Use "mailbox" topic approach for friend requests

### LINKED ISSUE #74: Create a global inbox topic for the user
All the client applications should subscribe to a topic that is specific for the peer itself:
eg. `/peer/<DID>`

Messages that are directed to the given user should be sent over that pubsub topic (eg. friend requests)

### PR DESCRIPTION
**Pull request by iltumio**
**What this PR does** 📖
Uses a "mailbox" topic approach for friend requests

**Which issue(s) this PR fixes** 🔨
Resolves #74

**Special notes for reviewers** 🗒️
Friend requests just works as before, but they are not using the FRIENDS_BROADCAST channel

**Additional comments** 🎤

### CODE REVIEW

File: extensions/warp-mp-ipfs/src/store/friends.rs [hunk 1]
```diff
@@ -385,21 +386,28 @@ impl<T: IpfsTypes> FriendsStore<T> {
...
+            let Queue(peer, data, recipient) = item;
             if let Ok(crate::store::PeerConnectionType::Connected) =
                 connected_to_peer(self.ipfs.clone(), *peer).await
             {
                 let bytes = serde_json::to_vec(&data)?;
 
-                self.ipfs
-                    .pubsub_publish(FRIENDS_BROADCAST.into(), bytes)
-                    .await?;
+                let sender = data.clone().sender().unwrap();
```
Comment (dariusc93) [inline extensions/warp-mp-ipfs/src/store/friends.rs:395]:
```suggestion
                let sender = data.clone().sender().ok_or(Error::Other)?;
```

File: extensions/warp-mp-ipfs/src/store/mod.rs [hunk 2]
```diff
@@ -26,6 +26,10 @@ pub const IDENTITY_BROADCAST: &str = "identity/broadcast";
 pub const FRIENDS_BROADCAST: &str = "friends/broadcast";
 pub const SYNC_BROADCAST: &str = "/identity/sync/broadcast";
 
+pub fn get_inbox_topic(did: &DID) -> String {
```
Comment (dariusc93) [inline extensions/warp-mp-ipfs/src/store/mod.rs:29]: this probably should go into the friends.rs file

File: extensions/warp-mp-ipfs/src/store/friends.rs [hunk 3]
```diff
@@ -155,7 +156,7 @@ impl<T: IpfsTypes> FriendsStore<T> {
         let store = Self {
             ipfs,
             identity,
-            did_key,
+            did_key: did_key.clone(),
```
Comment (dariusc93) [inline extensions/warp-mp-ipfs/src/store/friends.rs:159]: cloning shouldnt be when you can use it from `store.did_key`

File: extensions/warp-mp-ipfs/src/store/friends.rs [hunk 4]
```diff
@@ -385,21 +386,28 @@ impl<T: IpfsTypes> FriendsStore<T> {
...
             {
                 let bytes = serde_json::to_vec(&data)?;
 
-                self.ipfs
-                    .pubsub_publish(FRIENDS_BROADCAST.into(), bytes)
-                    .await?;
+                let sender = data.clone().sender().unwrap();
+
+                let recipients = match data.recipients() {
+                    Some(rec_vec) => {rec_vec.into_iter().filter(move |rec| *rec.fingerprint() != sender.fingerprint() ).collect::<Vec<KeyPair>>()},
```
Comment (dariusc93) [inline extensions/warp-mp-ipfs/src/store/friends.rs:398]: we could simply this by doing something like

```rust
data.recipients().map(|list| list.iter().filter(move |rec| *rec.fingerprint() != sender.fingerprint() ).collect::<Vec<_>>()).unwrap_or_default()
```

though i dont see this being used anyplace and probably wouldnt be needed

**Review by dariusc93 — COMMENTED** [review]
I did leave a couple of comments

File: extensions/warp-mp-ipfs/src/store/mod.rs (hunk 2 above)
Comment (iltumio) [reply extensions/warp-mp-ipfs/src/store/mod.rs:29]: ok, we can think of a generalization of that concept in the future

File: extensions/warp-mp-ipfs/src/store/friends.rs (hunk 3 above)
Comment (iltumio) [reply extensions/warp-mp-ipfs/src/store/friends.rs:158]: ok, done

**Comment by dariusc93** [comment]
I will be doing some side testing before merging this in to confirm a few things internally.

**Comment by dariusc93** [comment]
Will still continue doing some test on this but will merge this as it is now but may change in the future using the same approach

**Pull request merged**